test/integration/pilot_test.go

```diff
@@ -0,0 +1,200 @@
+package integration_test
+
+import (
+	"os"
+	"os/exec"
+	"path/filepath"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/onsi/gomega/gbytes"
+	"github.com/pkg/errors"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"sigs.k8s.io/testing_frameworks/integration"
+
+	"github.com/jetstack/navigator/internal/test/util/testfs"
+)
+
+// MakeKubeConfig creates a kube/config file which is suitable for communicating
+// with the API server started by ``integration.ControlPlane``.
+func MakeKubeConfig(f *integration.ControlPlane, path string) error {
+	ctl := f.KubeCtl()
+	ctl.Opts = []string{"--kubeconfig", path}
+
+	_, _, err := ctl.Run(
+		"config",
+		"set-credentials",
+		"user1",
+	)
+	if err != nil {
+		return errors.Wrap(err, "unable to create user")
+	}
+
+	_, _, err = ctl.Run(
+		"config",
+		"set-cluster",
+		"integration1",
+		"--server",
+		f.APIURL().String(),
+		"--certificate-authority",
+		filepath.Join(f.APIServer.CertDir, "apiserver.crt"),
+	)
+	if err != nil {
+		return errors.Wrap(err, "unable to create cluster")
+	}
+
+	_, _, err = ctl.Run(
+		"config",
+		"set-context",
+		"default",
+		"--cluster", "integration1",
+		"--user", "user1",
+	)
+	if err != nil {
+		return errors.Wrap(err, "unable to create context")
+	}
+
+	_, _, err = ctl.Run(
+		"config",
+		"use-context",
+		"default",
+	)
+	if err != nil {
+		return errors.Wrap(err, "unable to use context")
+	}
+	return nil
+}
+
+func TestPilotCassandra(t *testing.T) {
+	// Start the API server with CustomResourceSubresources feature enabled.
+	// Navigator pilot calls on UpdateStatus on Pilot resources for example.
+	cp := &integration.ControlPlane{
+		APIServer: &integration.APIServer{
+			Args: []string{
+				"--etcd-servers={{ if .EtcdURL }}{{ .EtcdURL.String }}{{ end }}",
+				"--cert-dir={{ .CertDir }}",
+				"--insecure-port={{ if .URL }}{{ .URL.Port }}{{ end }}",
+				"--insecure-bind-address={{ if .URL }}{{ .URL.Hostname }}{{ end }}",
+				"--secure-port=0",
+				"--feature-gates=CustomResourceSubresources=true",
+				"-v=4", "--alsologtostderr",
+			},
+			// Out: os.Stdout,
+			// Err: os.Stderr,
+		},
+	}
+	err := cp.Start()
+	require.NoError(t, err)
+	defer func() {
+		err := cp.Stop()
+		if err != nil {
+			t.Fatal(err)
+		}
+	}()
+	cli := cp.KubeCtl()
+
+	// Create a CRD for the Navigator Pilot resource type.
+	// This allows us to test the Pilot against the Kubernetes API server,
+	// without having to start the Navigator API server and configure aggregation etc.
+	crdPath, err := filepath.Abs("testdata/pilot_crd.yaml")
+	require.NoError(t, err)
+	stdout, stderr, err := cli.Run(
+		"apply",
+		"--filename",
+		crdPath,
+	)
+	t.Log("stdout", stdout)
+	t.Log("stderr", stderr)
+	require.NoError(t, err)
```

Comment:
From here up to the start of the function should be in some shared function - this is code that is not specific to this test, or Cassandra.

Also, we should add some comments to this explaining *why* we are registering a CRD, and what is and isn't okay to test whilst using a CRD (because a lot of code for e.g. validation won't be run, we should clearly define bounds for what is and isn't okay to test). As we spoke about before, I think we should look to remove this code ASAP in favour of launching navigator-apiserver itself.

Also, given what we're trying to test here, would it be possible to implement this whole test as a unit test? I'm not sure why we need to bring up the entire process to verify that a config file is written ❓